lua/nvim-tree/async.lua

```diff
@@ -0,0 +1,180 @@
+---Idea taken from: https://github.com/ms-jpq/lua-async-await
+local co = coroutine
+
+local M = {}
+
+---@type table<thread, boolean>
+local async_threads = {}
+
+---Execuate an asynchronous function
+---@param func function
+---@param ... any The arguments passed to `func`, plus a callback receives (err, ...result)
+function M.exec(func, ...)
+  local nargs = select("#", ...)
+  local args = { ... }
+  ---@type function
+  local cb = function() end
+  if nargs > 0 then
+    cb = args[nargs]
+    args[nargs] = nil
+  end
+
+  local thread = co.create(func)
+  async_threads[thread] = true
+
+  local function step(...)
+    local res = { co.resume(thread, ...) }
+    local ok = table.remove(res, 1)
+    local err_or_next = res[1]
+    if co.status(thread) ~= "dead" then
+      local _, err = xpcall(err_or_next, debug.traceback, step)
+      if err then
+        cb(err)
+      end
+    else
+      async_threads[thread] = nil
+      if ok then
+        cb(nil, unpack(res))
+      else
+        cb(debug.traceback(thread, err_or_next))
+      end
+    end
+  end
+
+  step(unpack(args))
+end
+
+---Test whether we are in async context
+---@return boolean
+function M.in_async()
+  local thread = co.running()
+  return async_threads[thread] ~= nil
+end
+
+---Wrap an asynchronous function to be directly called in synchronous context
+---@param func function
+---@param argc number The number of arguments the wrapped function accepts. Pass it if you want the returned function to receive an additional callback as final argument, and the signature of callback is the same as that of `async.exec`.
+---@return function
+function M.wrap(func, argc)
+  return function(...)
+    local args = { ... }
+    if argc == nil or #args == argc then
+      table.insert(args, function() end)
+    end
+    M.exec(func, unpack(args))
+  end
+end
+
+---Asynchronously call a function, which has callback as the last parameter (like luv apis)
+---@param func function
+---@param ... any
+---@return any
+function M.call(func, ...)
+  local args = { ... }
+  return co.yield(function(cb)
+    table.insert(args, cb)
+    func(unpack(args))
+  end)
+end
+
+---Execuate multiple asynchronous function simultaneously
+---@param ... fun()
+---@return table[] (err, ...result) tuples from every function
+function M.all(...)
+  local tasks = { ... }
+  if #tasks == 0 then
+    return {}
+  end
+
+  local results = {}
+  local finished = 0
+  return co.yield(function(cb)
+    for i, task in ipairs(tasks) do
+      M.exec(task, function(...)
+        finished = finished + 1
+        results[i] = { ... }
+        if finished == #tasks then
+          cb(results)
+        end
+      end)
+    end
+  end)
+end
+
+---Asynchronous `vim.schedule`
```

Comment:
As documented in `:h lua-loop-callbacks`, we cannot call some `vim.api.*` and all `vim.fn.*` in luv callbacks. So every time before these calls, we should manually call `vim.schedule` (`async.schedule` here to avoid callback hells). It's my fault to assume that we already get this; I'll add necessary comments about it.

Reply:
I see. So this schedule is essentially a thread join.

I do not understand the timing; surely it should be called immediately after the `.call` or during the callback. It's called once at the start of `create_file` and once at the end.

Reply:
Certainly we could do that for every callback, but usually this is only needed when the callback contains vim apis (or we are uncertain about this, such as user provided callback). In some cases we may not want to use `vim.schedule` to delay the execution of callback like a sequence of consecutive luv calls and always delaying may have potential negative impact on performance.

Reply:
I see. In that case, please move the calls to schedule immediately after the calls that require this join. It should be clear exactly when and where it is required.